runtime/src/account_saver.rs

```diff
@@ -110,18 +110,19 @@ fn collect_accounts_for_successful_tx<'a, T: SVMMessage>(
     transaction_ref: Option<&'a SanitizedTransaction>,
     transaction_accounts: &'a [TransactionAccount],
 ) {
-    for (_, (address, account)) in (0..transaction.account_keys().len())
-        .zip(transaction_accounts)
-        .filter(|(i, _)| {
-            transaction.is_writable(*i) && {
-                // Accounts that are invoked and also not passed as an instruction
-                // account to a program don't need to be stored because it's assumed
-                // to be impossible for a committable transaction to modify an
-                // invoked account if said account isn't passed to some program.
-                !transaction.is_invoked(*i) || transaction.is_instruction_account(*i)
-            }
-        })
-    {
+    for (i, (address, account)) in transaction_accounts.iter().enumerate() {
```

Comment:
You need to limit the iteration to `transaction.account_keys().len()` because `transaction_accounts` includes some appended program accounts.

Reply:
youre right, thank you for catching that

unrelated to this pr, but do you know a reason i might have missed re: why we would need to append them? the extra loaders added in `load_transaction_accounts()`, which i believe is the only place accounts are added. i forgot we do it because i havent been doing it in my new account loader, and it doesnt seem to cause any problems in testing. `ProgramCacheForTxBatch` is populated with all builtins and as best i can tell invoke context always gets loaders from there

(to be clear i still count them against data size limits)

Reply:
I'm not exactly sure at this point whether it's necessary to append them or not. The program cache came later so it's very possible it alleviates the need to do this unintuitive appending behavior.